illumos-utils/src/destructor.rs

```diff
@@ -0,0 +1,389 @@
+                    };
+
+                    match msg {
+                        Message::Data(object) => self.enqueue_destroy(object),
+                        Message::Exit => {
+                            exit = true;
+                            if self.futs.is_empty() {
+                                return;
+                            }
+                        },
+                    }
+                },
+            }
+        }
+    }
+
+    fn enqueue_destroy(&self, object: T) {
+        self.futs.push(ShutdownWaitFuture(
+            async move {
+                let do_delete = || async {
+                    object.delete().await.map_err(|err| {
+                        BackoffError::transient(err.to_string())
+                    })?;
+                    Ok(())
+                };
+                let log_failure = |_err: String, _| {};
+                retry_notify(
+                    retry_policy_internal_service(),
+                    do_delete,
+                    log_failure,
+                )
+                .await?;
+                Ok(())
+            }
+            .boxed()
+            .shared(),
+        ));
+    }
+}
+
+struct Inner<T> {
+    tx: mpsc::UnboundedSender<Message<T>>,
+    handle: Option<tokio::task::JoinHandle<()>>,
+}
+
+/// A destructor which asynchronously destroys objects,
+/// which can throw errors during destruction.
+pub struct Destructor<T> {
+    inner: Arc<Mutex<Inner<T>>>,
+}
+
+// I wish this could be derived, but the derive(Clone) macro is silly,
+// and has the impression that "T: Clone", which is not true.
+impl<T> Clone for Destructor<T> {
+    fn clone(&self) -> Self {
+        Self { inner: self.inner.clone() }
+    }
+}
+
+impl<T: Deletable> Destructor<T> {
+    /// Creates a new destructor with a background task to consume destroyed
+    /// objects.
+    pub fn new() -> Self {
+        let futs = FuturesUnordered::new();
+        let (tx, rx) = mpsc::unbounded_channel::<Message<T>>();
+
+        let handle = tokio::task::spawn(async move {
+            let mut worker = DestructorWorker { rx, futs };
+            worker.run().await;
+        });
+        Self { inner: Arc::new(Mutex::new(Inner { tx, handle: Some(handle) })) }
+    }
+
+    /// Destroys an object asynchronously in a background task.
+    pub fn enqueue_destroy(&self, object: T) {
+        self.inner
+            .lock()
+            .unwrap()
+            .tx
+            .send(Message::Data(object))
+            .map_err(|err| err.to_string())
+            // Unwrap safety:
+            // - The DestructorWorker can only be stopped by "try_close"
+            // - "try_close" can only succeed if it's called on the last strong reference to the
+            // Destructor.
+            // - Since "enqueue_destroy" was called on a "Destructor", one more reference must
+            // exist.
+            .unwrap();
+    }
+
+    /// Closes the destructor if this is the only reference to it.
```

Comment:
Maybe make a note that `Self` is returned in the `Result::Err` if there is more than one reference outstanding.